cmd/private-org-sync/main.go

```diff
@@ -398,6 +392,89 @@ func (g gitSyncer) mirror(repoDir string, src, dst location) error {
 	return nil
 }
 
+func addGitRemote(logger *logrus.Entry, git gitFunc, org, repo, repoDir, remoteName string) error {
+	remoteSetupLogger := logger.WithField("remote-name", remoteName)
+	remoteSetupLogger.Debug("Remote does not exist, setting up")
+
+	srcUrl, err := url.Parse(fmt.Sprintf("https://github.com/%s/%s", org, repo))
+	if err != nil {
+		remoteSetupLogger.WithError(err).Error("Failed to construct URL for the source remote")
+		return fmt.Errorf("failed to construct URL for the source remote")
+	}
+
+	remoteSetupLogger = remoteSetupLogger.WithField("remote-url", srcUrl.String())
+	remoteSetupLogger.Debug("Adding remote")
+
+	if _, exitCode, err := git(logger, repoDir, "remote", "add", remoteName, srcUrl.String()); err != nil || exitCode != 0 {
+		remoteSetupLogger.WithField("exit-code", exitCode).WithError(err).Error("Failed to set up source remote")
+		return fmt.Errorf("failed to set up source remote")
+	}
+
+	return nil
+}
+
+func mergeRemotesAndPush(logger *logrus.Entry, git gitFunc, repoDir, srcRemote, srcOrg, srcRepo, destRemote, org, repo, branch, destUrl string, confirm bool) error {
+	workDir := filepath.Join(repoDir, "mergeSourceDest")
+
+	logger.Debugf("Creating new working directory: %s", workDir)
+	if err := os.MkdirAll(workDir, 0755); err != nil {
+		return fmt.Errorf("couldn't create working directory: %v", err)
+	}
+
+	logger.Debug("Initializing git repository")
+	if _, exitCode, err := git(logger, workDir, "init"); err != nil || exitCode != 0 {
+		logger.WithField("exit-code", exitCode).WithError(err).Error("Failed to initialize local git directory")
+		return fmt.Errorf("failed to initialize local git directory")
+	}
+
+	if err := addGitRemote(logger, git, srcOrg, srcRepo, workDir, srcRemote); err != nil {
+		return err
+	}
```

Comment:
Quick note: we definitely should not need to create a new working directory and fetch the source. At this stage we already have a working directory with the source fully fetched, all we need to do is to fetch the destination, merge the source to it and push to destination again. This looks like it does more work that is actually needed.

Reply:
Unfortunately the problem is that we have a source fully fetched under a bare git repo. I couldn't make that work without breaking the current design.

Reply:
Can we use a non-bare working repo from the start? Until the merge we are only doing fetches and pushes which should behave identically.

Reply:
I tried that, and the whole fetch a depth logic wasn't working. It also generated an empty git folder.  To be honest, I didn't spend much time debugging why, because I didn't intend to change the current design. However, let me try to find a way how to reuse the existing fetched source.

Reply:
Interesting. At the very worst, we could clone the new non-bare repository from the bare one (that's at least a local operation), but I think we should make it work with a single repo (which needs to be non-bare to allow merges).